### PR TITLE
fix: match npm install semantics by default

### PR DESCRIPTION
**Pull request by Bubblyworld**
Changes behaviour with existing locks to match `npm` by default (i.e. if other
install flags like `freeze` aren't set). The only difference is how we handle
secondary locks: if they're in range for their parent `package.json`, we now
use the lock, and if they're out of range we reset them to latest compatible,
which is what `npm` does in these cases.

### CODE REVIEW

**Comment by Bubblyworld** [comment]
Note that these lock changes also affect non-install operations, which again matches `npm`'s behaviour.

File: src/install/installer.ts [hunk 1]
```diff
@@ -363,14 +363,55 @@ export class Installer {
...
+      (isTopLevel && pcfg.devDependencies?.[pkgName]);
+    const pjsonTarget =
+      pjsonTargetStr &&
+      newPackageTarget(
+        pjsonTargetStr,
+        new URL(definitelyPkgScope),
+        this.defaultRegistry,
+        pkgName
+      );
+    const resolutionInRange = async (resolution: InstalledResolution) => {
```
Comment (Bubblyworld) [inline src/install/installer.ts:386]: This function checks that an existing resolution is in-range for the parent scope's `package.json` (if it has one).

File: src/install/installer.ts [hunk 2]
```diff
@@ -380,20 +421,21 @@ export class Installer {
...
 
-    // flattened resolution cascading for secondary
-    if (
-      (!isTopLevel && opts.mode.includes("existing") && !opts.latest) ||
-      (!isTopLevel && opts.mode.includes("new") && opts.freeze)
-    ) {
+    // Pick up resolutions from flattened scopes like 'https://ga.jspm.io/"
+    // for secondary installs, if they're in range for the current pjson, or
+    // if we're in a freeze install:
+    if (!isTopLevel) {
```
Comment (Bubblyworld) [inline src/install/installer.ts:427]: Same logic but for flattened resolutions instead.

File: test/api/extract.test.js [hunk 3]
```diff
@@ -29,7 +29,7 @@ assert.strictEqual(
 );
 assert.strictEqual(
   map.scopes["https://ga.jspm.io/"]["object-assign"],
-  "https://ga.jspm.io/npm:object-assign@4.1.0/index.js"
+  "https://ga.jspm.io/npm:object-assign@4.1.1/index.js"
```
Comment (Bubblyworld) [inline test/api/extract.test.js:32]: A lot of these test changes are from secondaries being bumped to latest compatible.

File: test/html/inject.test.js [hunk 4]
```diff
@@ -55,7 +55,7 @@ assert.strictEqual(
       '<script type="importmap">\n' +
       "{\n" +
       '  "imports": {\n' +
-      '    "react": "https://ga.jspm.io/npm:react@17.0.2/index.js"\n' +
+      '    "react": "https://ga.jspm.io/npm:react@16.14.0/index.js"\n' +
```
Comment (Bubblyworld) [inline test/html/inject.test.js:58]: Latest compatible in this generator's base is 16.14.0.

File: src/install/installer.ts (hunk 1 above)
Comment (guybedford) [reply src/install/installer.ts:386]: We are supposed to already have this function from the `inRange` function on the class itself - can we reuse / refactor into the existing implementation rather?

File: src/install/installer.ts [hunk 5]
```diff
@@ -363,14 +363,55 @@ export class Installer {
...
+    // Find any existing locks in the current package scope, making sure
+    // secondaries are always in-range for their parent scope pjsons:
     const existingResolution = getResolution(
       this.installs,
       pkgName,
       isTopLevel ? null : pkgScope
     );
-    if (existingResolution) {
+    if (
+      existingResolution &&
```
[the remark this answers is not in the document]
Comment (guybedford) [reply src/install/installer.ts:410]: Perfect

File: test/api/extract.test.js (hunk 3 above)
Comment (guybedford) [reply test/api/extract.test.js:32]: Could we add `freeze` to the test case? That will also help avoid new releases shifting our tests which is a common issue.

File: test/html/inject.test.js (hunk 4 above)
Comment (guybedford) [reply test/html/inject.test.js:58]: I don't see `react` in the base `package.json` file?

**Review by guybedford — COMMENTED** [review]
Excellent, thank you for nailing this down that's really important.

File: test/html/inject.test.js (hunk 4 above)
Comment (Bubblyworld) [reply test/html/inject.test.js:58]: Ah, the base url here is `/test/html/local` (see line 6), which has the following `package.json`:

```
{
  "dependencies": {
    "react": "^16"
  }
}
```

File: src/install/installer.ts (hunk 1 above)
Comment (Bubblyworld) [reply src/install/installer.ts:386]: Ah, right, I didn't see that - can do 👌

**Pull request merged**